src/vm/comwaithandle.cpp

```diff
@@ -157,99 +126,66 @@ INT64 AdditionalWait(INT64 sPauseTime, INT64 sTime, INT64 expDuration)
     return additional;
 }
 
-FCIMPL3(INT32, WaitHandleNative::CorWaitOneNative, SafeHandle* safeWaitHandleUNSAFE, INT32 timeout, CLR_BOOL exitContext)
+FCIMPL2(INT32, WaitHandleNative::CorWaitOneNative, HANDLE handle, INT32 timeout)
 {
     FCALL_CONTRACT;
 
     INT32 retVal = 0;
-    SAFEHANDLEREF sh(safeWaitHandleUNSAFE);
-    HELPER_METHOD_FRAME_BEGIN_RET_1(sh);
+    HELPER_METHOD_FRAME_BEGIN_RET_0();
 
-    _ASSERTE(sh != NULL);
+    _ASSERTE(handle != 0);
+    _ASSERTE(handle != INVALID_HANDLE_VALUE);
 
     Thread* pThread = GET_THREAD();
 
     DWORD res = (DWORD) -1;
 
-    SafeHandleHolder shh(&sh);
-    // Note that SafeHandle is a GC object, and RequestCallback and 
-    // DoAppropriateWait work on an array of handles.  Don't pass the address
-    // of the handle field - that's a GC hole.  Instead, pass this temp
-    // array.
-    HANDLE handles[1];
-    handles[0] = sh->GetHandle();
+    // Support for pause/resume (FXFREEZE)
+    while(true)
     {
-        // Support for pause/resume (FXFREEZE)
-        while(true)
-        {
-            INT64 sPauseTime = g_PauseTime;
-            INT64 sTime = CLRGetTickCount64();
-            res = pThread->DoAppropriateWait(1,handles,TRUE,timeout, WaitMode_Alertable /*alertable*/);
-            if(res != WAIT_TIMEOUT)
-                break;
-            timeout = (INT32)AdditionalWait(sPauseTime, sTime, timeout);
-            if(timeout == 0)
-                break;
-        }
+        INT64 sPauseTime = g_PauseTime;
+        INT64 sTime = CLRGetTickCount64();
+        res = pThread->DoAppropriateWait(1, &handle, TRUE, timeout, (WaitMode)(WaitMode_Alertable | WaitMode_IgnoreSyncCtx));
```

Comment:
Ah, this is a valid concern. I moved the pause time adjustment logic into the original `WaitHelper` method. If someone overrides the `SynchronizationContext` and calls back into the `base` method it will still do the loop and in a way that is more "invisible" to the managed code. However if someone overrides it and decides to implement some wait logic themselves without calling into the `base` methods it will become a problem.

Reply:
I looked up Google/GitHub/ReferenceSource for some implementations where the change would break something and came up with this:

https://referencesource.microsoft.com/#WindowsBase/Base/System/Windows/Threading/DispatcherSynchronizationContext.cs,3d51d8610abbe8d8

Remaining usages would be unaffected by the change (copied CoreFX, Mono, Reference Source, SSCLI, XNA, etc. code; code that eventually called `base.Wait` or `WaitHelper`).

Reply:
I looked into it a bit and I would probably be able to expose the pause time as a static field and then set it from unmanaged code using `MscorlibBinder::GetField(...)->SetStaticValue64(...)`. That way it will only have some performance hit when the Pause/Resume actually happens and the retry loop could live in the managed code.

Reply:
On a second thought it's not such a great idea to create a field that is accessed from unmanaged code. The `AdditionalWait` calculation requires `TickCount` anyway, so I won't be able to avoid a call into the runtime.

Reply:
You can delete `g_PauseTime`, `CorExecutionManager::Pause/Resume` and everything related to it. This was a feature built for Windows Phone that is not needed anymore.

Reply:
I'll be happy to do so in separate PR.

Reply:
Preview here: https://github.com/dotnet/coreclr/commit/58c2b5f7047b90bca1511ef158da7a113da18edc

Reply:
Looks good to me

Reply:
Good, I will submit it once this one is merged.